### PR TITLE
Build: Update `npm run dev` watch task to a single call to watch function

### PR DESCRIPTION
**Pull request by talldan**
## Description
This PR was an attempt to improve the performance of `npm run dev`. It doesn't seem to make much of a difference, but at least tidies up the code, and switches to using the node-watch API in a way that's closer to the author's intention (replacing the multiple calls to `watch` for each package with a single call that uses the `filter` property).

Also fixes a bug where the watcher wasn't deleting build files, the previous code was creating an incorrect path to the file to be deleted. It now works in this PR.

## How has this been tested?
1. Run `npm run dev`
2. Change some source code - the code should rebuild
3. Change some non-source code (e.g. test files) - the code should not rebuild
4. Delete a file - the deletion should be logged in the console and the corresponding build file removed.

### CODE REVIEW

File: bin/packages/watch.js [hunk 1]
```diff
@@ -26,63 +25,55 @@ const exists = ( filename ) => {
...
 		! [
-			/\/(benchmark|__mocks__|__tests__|test|storybook|stories)\/.+.js$/,
+			/\/(benchmark|__mocks__|__tests__|test|storybook|stories)\/.+/,
 			/.\.(spec|test)\.js$/,
-		].some( ( regex ) => regex.test( filename ) ) &&
-		/.\.(js|json|scss)$/.test( filename )
-	);
+		].some( ( regex ) => regex.test( filename ) );
+
+	return isSource ? true : skip;
```
Comment (talldan) [inline bin/packages/watch.js]: This should be a nice improvement, as `skip` avoids watching non-matching subfolders entirely.

File: bin/packages/watch.js [hunk 2]
```diff
@@ -26,63 +25,55 @@ const exists = ( filename ) => {
...
+	( event, filename ) => {
+		const filePath = path.resolve( PACKAGES_DIR, filename );
+		const srcDir = getSrcDirectory( filePath );
+		if ( event === 'update' && exists( filePath ) ) {
+			console.log( chalk.green( '->' ), `${ event }: ${ filePath }` );
+			rebuild( filePath );
+		} else {
+			const buildFile = path.resolve( srcDir, '..', 'build', filename );
+			try {
+				fs.unlinkSync( buildFile );
```
Comment (talldan) [inline bin/packages/watch.js]: This wasn't actually working in `trunk`. Any errors were being swallowed. The issue is fixed now.

**Comment by talldan** [comment]
> Have you tested how much it improved after applying these changes? Maybe some kind of a naive audit? Would be nice to include it in the description too :)

Only by checking the activity monitor. Any ideas how I could test it?

**Comment by kevin940726** [comment]
> Only by checking the activity monitor. Any ideas how I could test it?

I would imagine just something like a before&after of the time it takes from editing a file to finish building it? Or maybe CPU/memory usage?

**Comment by talldan** [comment]
I'll see if I can do some kind of profiling.

**Comment by talldan** [comment]
From what I can tell it doesn't really affect performance much, but it does at least tidy the code up and make fix the issue with deleting build files.

File: bin/packages/watch.js [hunk 3]
```diff
@@ -11,79 +11,154 @@ const chalk = require( 'chalk' );
...
+ * (usually triggered the associated source file was deleted)
+ *
+ * @param {'remove'} event The event name
+ * @param {string} filename
+ */
+function removeBuildFile( event, filename ) {
+	const buildFile = getBuildFile( filename );
+	if ( exists( buildFile ) ) {
+		try {
+			fs.unlinkSync( buildFile );
```
Comment (kevin940726) [inline bin/packages/watch.js]: Could probably use the async version of this method? I don't think it makes any difference in this context but could unblock the event loop when there're a lot of files being deleted at the same time (checking out a older branch for instance).

**Review by kevin940726 — COMMENTED** [review]
LGTM 👍! Just a tiny nitpick.

File: bin/packages/watch.js (hunk 3 above)
Comment (talldan) [reply bin/packages/watch.js]: Good suggestion 👍

**Comment by gziolo** [comment]
@talldan, what's blocking merging this PR?

**Comment by talldan** [comment]
@gziolo Just me taking my time to apply @kevin940726's suggestion. That's done now, hopefully this should be good to go.

**Comment by gziolo** [comment]
Cool, let's get this in then and check it in action 👍🏻

**Comment by talldan** [comment]
I'm going to merge then. 👍 

If you spot anything not working, feel free to revert and then I can follow up with a fix.

**Comment by gziolo** [comment]
Thank you for working on improvements to the build process 💯

**Comment by aristath** [comment]
[Discussed this on Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1618998688045300) but posting here for better tracking: This PR broke watching on Ubuntu. The first build runs, and then there are no builds on subsequent changes.

**Comment by vdwijngaert** [comment]
> [Discussed this on Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1618998688045300) but posting here for better tracking: This PR broke watching on Ubuntu. The first build runs, and then there are no builds on subsequent changes.

Looks like the [watcher that's used here](https://github.com/yuanchuan/node-watch) has some issues with recursive watching on linux. I'll try and see if I can make it work.

**Pull request merged**